eisbuk-admin/functions/src/https.ts

```diff
@@ -61,104 +59,127 @@ export const sendEmail = functions
     return { ...payload, success: true };
   });
 
-type Protocol = "https" | "http";
-
 /**
  * Sends SMS message using template data from organizations firestore entry and provided params
  */
 export const sendSMS = functions
   .region("europe-west6")
-  .https.onCall(async (payload: Partial<SendSMSPayload>, { auth }) => {
-    // check payload
-    if (!payload) {
-      throw new functions.https.HttpsError(
-        "invalid-argument",
-        HTTPErrors.NoPayload
+  .https.onCall(
+    async ({ organization, to, message }: SendSMSPayload, { auth }) => {
+      await checkUser(organization, auth);
+
+      // get SMS template data
+      const orgData =
+        ((
+          await admin
+            .firestore()
+            .collection(Collection.Organizations)
+            .doc(organization!)
+            .get()
+        ).data() as OrganizationData) || {};
+
+      const smsUrl = orgData.smsUrl!;
+      // Non numeric senders must be 11 chars or less
+      const smsSender = (orgData.smsSender || organization!)
+        .toString()
+        .substring(0, 11);
+      functions.logger.log(`Got smsSender: ${smsSender}, smsUrl: ${smsUrl}`);
+
+      // get sms secrets
+      const { smsAuthToken: authToken } =
+        (
+          await admin
+            .firestore()
+            .collection(Collection.Secrets)
+            .doc(organization!)
+            .get()
+        ).data() || ({} as OrganizationSecrets);
+
+      const { proto, ...options } = createSMSReqOptions(
+        "POST",
+        smsUrl,
+        authToken
       );
-    }
-
-    const { organization, to, message } = payload;
 
-    await checkUser(organization, auth);
-
-    // check payload validity
-    if (!to) {
-      throw new functions.https.HttpsError(
-        "invalid-argument",
-        SendSMSErrors.NoRecipient
-      );
-    }
-    if (!message) {
-      throw new functions.https.HttpsError(
-        "invalid-argument",
-        SendSMSErrors.NoMsgBody
+      const data = {
+        message,
+        // if sender not provided, fall back to organization name
+        sender: smsSender,
+        recipients: [{ msisdn: to }],
+      };
+      functions.logger.log("Sending POST data:", data);
+
+      const res = await sendRequest<SMSResponse>(proto, options, data);
+
+      if (res && res.ids) {
+        // A response containg a `res` key is successful
+        functions.logger.log("SMS POST request successfull", { response: res });
+      } else {
+        functions.logger.error("Error with SMS POST", { response: res });
+        // if res unsuccessful, throw
+        throw new functions.https.HttpsError(
+          "cancelled",
+          SendSMSErrors.SendingFailed,
+          res
+        );
+      }
+
+      const smsId = res.ids[0];
+
+      const [smsOk, status, errorMessage] = await pRetry(
+        () => checkSMS(smsId, authToken),
+        { maxRetryTime: 5000 }
```

Comment:
I think we should use something else to achieve our goals.

Our goal is to make sure that, no matter what, our function will return in less than X seconds. Imagine, for instance, that the remote service is taking 60 seconds to process requests.

`maxRetryTime` [comes into play when it's time to retry something](https://github.com/tim-kos/node-retry/commit/339a30a8b957a88cd68a8e14bbcf7462b340b4ea). So in the example above, if the remote service always takes 60 seconds, this invocation will take 60 seconds, and not retry a second time (since we exceeded the five seconds allowance `maxRetryTime` gave us.

I think we need something like [`p-timeout`](https://www.npmjs.com/package/p-timeout) and/or the firebase [`timeoutSeconds`](https://firebase.google.com/docs/functions/manage-functions#set_timeout_and_memory_allocation) option to pass to `runWith`.